client/auth.go

```diff
@@ -64,50 +74,95 @@ func (c *Conn) readInitialHandshake() error {
 		// mysql-proxy also use 12
 		// which is not documented but seems to work.
 		c.salt = append(c.salt, data[pos:pos+12]...)
+		pos += 13
+		// auth plugin
+		if end := bytes.IndexByte(data[pos:], 0x00); end != -1 {
+			c.authPluginName = string(data[pos : pos+end])
+		} else {
+			c.authPluginName = string(data[pos:])
+		}
+	}
+
+	// if server gives no default auth plugin name, use a client default
+	if c.authPluginName == "" {
+		c.authPluginName = defaultAuthPluginName
 	}
 
 	return nil
 }
 
+// generate auth response data according to auth plugin
+func (c *Conn) genAuthResponse(authData []byte) ([]byte, bool, error) {
+	// password hashing
+	switch c.authPluginName {
+	case "mysql_native_password":
+		return CalcPassword(authData[:20], []byte(c.password)), false, nil
+	case "caching_sha2_password":
+		return CalcCachingSha2Password(authData, c.password), false, nil
+	case "sha256_password":
+		if len(c.password) == 0 {
+			return nil, true, nil
+		}
+		if c.tlsConfig != nil || c.proto == "unix" {
+			// write cleartext auth packet
```

Comment:
clear text?

Reply:
see here: https://dev.mysql.com/doc/refman/8.0/en/sha256-pluggable-authentication.html

```For clients that use the sha256_password plugin, passwords are never exposed as cleartext when connecting to the server. How password transmission occurs depends on whether a secure connection or RSA encryption is used:

If the connection is secure, an RSA key pair is unnecessary and is not used. This applies to encrypted connections that use TLS. The password is sent as cleartext but cannot be snooped because the connection is secure.

If the connection is not secure, and an RSA key pair is available, the connection remains unencrypted. This applies to unencrypted connections without TLS. RSA is used only for password exchange between client and server, to prevent password snooping. When the server receives the encrypted password, it decrypts it. A scramble is used in the encryption to prevent repeat attacks.

If a secure connection is not used and RSA encryption is not available, the connection attempt fails because the password cannot be sent without being exposed as cleartext.```

Reply:
I added ref as comment in the code